### PR TITLE
Add BinaryDocValuesField to replace BytesRef (ScriptDocValues)

### PR DESCRIPTION
**Pull request by jdconrad**
This change creates the classes required for the scripting fields API to provide a binary field composed of doc values using BytesRef as the representation returned to the user as a value.

### CODE REVIEW

**Review by rjernst — COMMENTED** [review]
I have a couple thoughts

**Review by rjernst — APPROVED** [review]
A couple thoughts, nothing blocking

**Review by stu-elastic — APPROVED** [review]
We can do the error messages separately.

However, please don't commit until the spotless changes are gone for files otherwise untouched by this PR.

**Comment by jdconrad** [comment]
@stu-elastic Made the previously described changes. Would appreciate you looking them over one more time please.

File: server/src/main/java/org/elasticsearch/script/field/DelegateDocValuesField.java [hunk 1]
```diff
@@ -11,13 +11,13 @@
...
 import java.io.IOException;
-import java.util.List;
+import java.util.Iterator;
 
 /**
  * A default {@link Field} to provide {@code ScriptDocValues} for fields
  * that are not supported by the script fields api.
  */
-public class DelegateDocValuesField implements DocValuesField {
+public class DelegateDocValuesField implements DocValuesField<Object> {
```
Comment (stu-elastic) [inline server/src/main/java/org/elasticsearch/script/field/DelegateDocValuesField.java:20]: Is this going to work for `Numeric`s like long fields?

**Review by stu-elastic — COMMENTED** [review]
Looks great.

Please ensure that `EmptyField` still works for numeric fields.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/60_script_doc_values_binary.yml [hunk 2]
```diff
@@ -20,7 +20,30 @@
             index: test
             id: 1
             body:
-                binary: U29tZSBiaW5hcnkgYmxvYg==
+                binary: "U29tZSBiaW5hcnkgYmxvYg=="
+
+    - do:
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/60_script_doc_values_binary.yml:25]: Can you add a test for the iterator?  Just add up the binary as ints or something.

Comment (stu-elastic) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/60_script_doc_values_binary.yml:25]: Ahh its line 76.

File: server/src/main/java/org/elasticsearch/script/field/DelegateDocValuesField.java (hunk 1 above)
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/script/field/DelegateDocValuesField.java:20]: Yes, ScriptDocValues type underneath is independent of the iterator type here which throws an UnsupportedOperationException.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/50_script_doc_values.yml [hunk 3]
```diff
@@ -440,3 +440,25 @@ setup:
...
+---
+"empty":
+  - do:
+      search:
+        rest_total_hits_as_int: true
+        body:
+          script_fields:
+            field:
+              script:
+                source: "int value = field('dne').get(1); value"
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/50_script_doc_values.yml:453]: Nice.

**Comment by jdconrad** [comment]
Added getters for EmptyField along w/ tests.

**Comment by jdconrad** [comment]
@stu-elastic @rjernst  Thanks for the many reviews!

**Pull request merged**